libvisual-plugins/plugins/input/alsa/input_alsa.c

```diff
@@ -167,14 +168,14 @@ int inp_alsa_init (VisPluginData *plugin)
 
        /* Setup a large buffer */
 
-	tmp = 1000000;
+	tmp = 1000 * 1000 / inp_alsa_var_samplerate * inp_alsa_var_frames_target;  // in micro seconds
```

Comment:
This is an integer division as all the operands are integers. The result is likely to be way off.

Reply:
@kaixiong I was aware of the integer aspect and took that into consideration when choosing the order of operations. When playing with it locally again, you still seem to have a point. In IPython:

```
In [10]: 1000 * 1000 // 44100 * 256
Out[10]: 5632

In [11]: 1000 * 1000 / 44100.0 * 256
Out[11]: 5804.988662131519

In [12]: 1000 * 1000 * 256 // 44100
Out[12]: 5804

In [13]: 1000 * 1000 * 256 // 44100 + 1
Out[13]: 5805
```

How about `1000 * 1000 * inp_alsa_var_frames_target / inp_alsa_var_samplerate + 1;`? `+1` would be for the fractional part that has been floored away.

Reply:
Let's also bracket the product for that extra clarity. +1 is good enough in my view. At worse we allocate 1 more sample/bytes than necessary when the division is perfect.

Is there a chance of overflow with that product? It looks like a pretty large number.

Reply:
> @hartwork, let's also bracket the product for that extra clarity.

@kaixiong no objections. Where do you want me to add braces?

> Is there a chance of overflow with that product? It looks like a pretty large number.

On 32bit `signed int tmp` will go up to `2 ** 30 - 1` i.e. `1073741823`. That allows a maximum value of 1073 for `inp_alsa_var_frames_target` before we overflow, or 1024 in power-of-two land:

```
In [4]: (2 ** 30 - 1) / 1_000_000
Out[4]: 1073.741823
```

How about this change:

```diff
--- static const int  inp_alsa_var_frames_target = 256;
+++ static const int  inp_alsa_var_frames_target = 256;  // must be <=1073 to not integer math overflow below
```

Reply:
@hartwork, it would just look like:
```
tmp = (1000 * 1000 * inp_alsa_var_frames_target) / inp_alsa_var_samplerate + 1
```

Getting nerdsniped by this issue. If we want to get the exact ceiling, I believe we can use this numerical trick:
```
tmp = (1000 * 1000 * inp_alsa_var_frames_target + inp_alsa_var_samplerate - 1) / inp_alsa_var_samplerate
```
Obviously this will reduce the effective range of `inp_alsa_var_samplerate`.

Reply:
@kaixiong maybe we need to collect requirements first. Minimum requirement would be no overflow and to not be off by more than 1. After that, simplicity and precision are in conflict in some sense, at least in pure integer world. With simplicity in mind, `+1` is already not clean and `+ inp_alsa_var_samplerate - 1` is even further from explaining itself, at least quickly. And both of these variants would be off by one if `inp_alsa_var_frames_target` is allowed to change freely. Maybe it's time to go for actual `ceil` and floating point to be clean, precise and self-explaining(?):
```
tmp = (int)ceil((1000.0 * 1000.0 * inp_alsa_var_frames_target) / inp_alsa_var_samplerate);
```
What do you think?

Reply:
Agreed, we might as well use floats here as it's not performance sensitive. It's clear and very hard to go wrong.

Reply:
@kaixiong okay great, now pushed with a fix.